source/s3_checksums.c

```diff
@@ -0,0 +1,270 @@
+        default:
+            return NULL;
+    }
+}
+const struct aws_byte_cursor *aws_get_create_mpu_header_name_from_algorithm(enum aws_s3_checksum_algorithm algorithm) {
+    switch (algorithm) {
+        case AWS_SCA_CRC32C:
+            return &g_crc32c_create_mpu_header_name;
+        case AWS_SCA_CRC32:
+            return &g_crc32_create_mpu_header_name;
+        case AWS_SCA_SHA1:
+            return &g_sha1_create_mpu_header_name;
+        case AWS_SCA_SHA256:
+            return &g_sha256_create_mpu_header_name;
+        default:
+            return NULL;
+    }
+}
+
+const struct aws_byte_cursor *aws_get_complete_mpu_name_from_algorithm(enum aws_s3_checksum_algorithm algorithm) {
+    switch (algorithm) {
+        case AWS_SCA_CRC32C:
+            return &g_crc32c_complete_mpu_name;
+        case AWS_SCA_CRC32:
+            return &g_crc32_complete_mpu_name;
+        case AWS_SCA_SHA1:
+            return &g_sha1_complete_mpu_name;
+        case AWS_SCA_SHA256:
+            return &g_sha256_complete_mpu_name;
+        default:
+            return NULL;
+    }
+}
+
+void s3_hash_destroy(struct aws_s3_checksum *checksum) {
+    struct aws_hash *hash = (struct aws_hash *)checksum->impl;
+    aws_hash_destroy(hash);
+    aws_mem_release(checksum->allocator, checksum);
+}
+
+int s3_hash_update(struct aws_s3_checksum *checksum, const struct aws_byte_cursor *to_checksum) {
+    struct aws_hash *hash = (struct aws_hash *)checksum->impl;
+    return aws_hash_update(hash, to_checksum);
+}
+
+int s3_hash_finalize(struct aws_s3_checksum *checksum, struct aws_byte_buf *output, size_t truncate_to) {
+    struct aws_hash *hash = (struct aws_hash *)checksum->impl;
+    checksum->good = false;
+    return aws_hash_finalize(hash, output, truncate_to);
+}
+
+typedef uint32_t (*crc_fn)(const uint8_t *, int, uint32_t);
+
+uint32_t aws_crc32_common(uint32_t previous, const struct aws_byte_cursor *buf, crc_fn checksum_fn) {
+
+    size_t length = buf->len;
+    uint8_t *buffer = buf->ptr;
+    uint32_t val = previous;
+    while (length > INT_MAX) {
+        val = checksum_fn(buffer, INT_MAX, val);
+        buffer += (size_t)INT_MAX;
+        length -= (size_t)INT_MAX;
+    }
+    return checksum_fn(buffer, (int)length, val);
+}
+
+int aws_crc_finalize(struct aws_s3_checksum *checksum, struct aws_byte_buf *out, size_t truncate_to) {
+    if (!checksum->good) {
+        return aws_raise_error(AWS_ERROR_INVALID_STATE);
+    }
+    checksum->good = false;
+    size_t available_buffer = out->capacity - out->len;
+    size_t len = checksum->digest_size;
+    if (truncate_to && truncate_to < len) {
+        len = truncate_to;
+    }
+    if (available_buffer < len) {
+        return aws_raise_error(AWS_ERROR_SHORT_BUFFER);
+    }
+    AWS_PRECONDITION(aws_byte_buf_is_valid(out));
+    uint32_t tmp = aws_hton32(*(uint32_t *)checksum->impl);
+    if (aws_byte_buf_write(out, (uint8_t *)&tmp, len)) {
+        return AWS_OP_SUCCESS;
+    }
+    return AWS_OP_ERR;
```

Comment:
you need raise error for this case, since `aws_byte_buf_write` will not raise error when it fails